File: frontend/app/services/pomodoroSessions.ts
import { ApiService } from './base/ApiService';

export interface PomodoroSession {
  id: number;
  user_id: number;
  date: string; // YYYY-MM-DD format
  duration: number; // in minutes
  type: 'work' | 'break';
  completed: boolean;
  completed_at: string;
  created_at: string;
  updated_at: string;
}

export interface CreatePomodoroSession {
  date: string;
  duration: number;
  type: 'work' | 'break';
  completed: boolean;
}

export interface UpdatePomodoroSession {
  date?: string;
  duration?: number;
  type?: 'work' | 'break';
  completed?: boolean;
}

export interface PomodoroStats {
  total_work_time_minutes: number;
}

export interface TodayWorkTimeStats {
  today_work_time_minutes: number;
}

class PomodoroSessionsService extends ApiService {
  async getSessions(): Promise<PomodoroSession[]> {
    return this.get<PomodoroSession[]>('/pomodoro-sessions/');
  }

  async getWeeklySessions(): Promise<PomodoroSession[]> {
    return this.get<PomodoroSession[]>('/pomodoro-sessions/weekly');
  }

  async getTodaySessions(): Promise<PomodoroSession[]> {
    return this.get<PomodoroSession[]>('/pomodoro-sessions/today');
  }

  async getTodayWorkSessions(): Promise<PomodoroSession[]> {
    return this.get<PomodoroSession[]>('/pomodoro-sessions/today/work');
  }

  async getTotalWorkTime(): Promise<PomodoroStats> {
    return this.get<PomodoroStats>('/pomodoro-sessions/stats/total-work-time');
  }

  async getTodayWorkTime(): Promise<TodayWorkTimeStats> {
    return this.get<TodayWorkTimeStats>('/pomodoro-sessions/stats/today-work-time');
  }

  async createSession(sessionData: CreatePomodoroSession): Promise<PomodoroSession> {
    return this.post<PomodoroSession>('/pomodoro-sessions/', sessionData);
  }

  async getSession(sessionId: number): Promise<PomodoroSession> {
    return this.get<PomodoroSession>(`/pomodoro-sessions/${sessionId}`);
  }

  async updateSession(sessionId: number, sessionData: UpdatePomodoroSession): Promise<PomodoroSession> {
    return this.put<PomodoroSession>(`/pomodoro-sessions/${sessionId}`, sessionData);
  }

  async deleteSession(sessionId: number): Promise<void> {
    return this.delete<void>(`/pomodoro-sessions/${sessionId}`);
  }
}

export const pomodoroSessionsService = new PomodoroSessionsService();

File: backend/app/routers/pomodoro_sessions.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from app.db.session import get_db
from app.api.deps import get_current_user
from app.models.user import User
from app.schemas.pomodoro_session import (
    CreatePomodoroSession, 
    UpdatePomodoroSession, 
    PomodoroSessionResponse
)
from app.services.pomodoro_sessions import PomodoroSessionService

router = APIRouter()

@router.get("/", response_model=List[PomodoroSessionResponse])
def get_pomodoro_sessions(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get all pomodoro sessions for the current user"""
    service = PomodoroSessionService(db)
    sessions = service.get_sessions(current_user.id, skip=skip, limit=limit)
    return sessions

@router.get("/weekly", response_model=List[PomodoroSessionResponse])
def get_weekly_sessions(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get pomodoro sessions for the current week"""
    service = PomodoroSessionService(db)
    sessions = service.get_weekly_sessions(current_user.id)
    return sessions

@router.get("/today", response_model=List[PomodoroSessionResponse])
def get_today_sessions(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get pomodoro sessions for today"""
    service = PomodoroSessionService(db)
    sessions = service.get_today_sessions(current_user.id)
    return sessions

@router.get("/today/work", response_model=List[PomodoroSessionResponse])
def get_today_work_sessions(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get completed work sessions for today"""
    service = PomodoroSessionService(db)
    sessions = service.get_today_work_sessions(current_user.id)
    return sessions

@router.get("/stats/total-work-time")
def get_total_work_time(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get total work time in minutes"""
    service = PomodoroSessionService(db)
    total_time = service.get_total_work_time(current_user.id)
    return {"total_work_time_minutes": total_time}

@router.get("/stats/today-work-time")
def get_today_work_time(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get today's work time in minutes"""
    service = PomodoroSessionService(db)
    today_time = service.get_today_work_time(current_user.id)
    return {"today_work_time_minutes": today_time}

@router.post("/", response_model=PomodoroSessionResponse)
def create_pomodoro_session(
    session_data: CreatePomodoroSession,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new pomodoro session"""
    service = PomodoroSessionService(db)
    session = service.create_session(current_user.id, session_data)
    return session

@router.get("/{session_id}", response_model=PomodoroSessionResponse)
def get_pomodoro_session(
    session_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get a specific pomodoro session"""
    service = PomodoroSessionService(db)
    session = service.get_session(current_user.id, session_id)
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Pomodoro session not found"
        )
    return session

@router.put("/{session_id}", response_model=PomodoroSessionResponse)
def update_pomodoro_session(
    session_id: int,
    session_data: UpdatePomodoroSession,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update a pomodoro session"""
    service = PomodoroSessionService(db)
    session = service.update_session(current_user.id, session_id, session_data)
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Pomodoro session not found"
        )
    return session

@router.delete("/{session_id}")
def delete_pomodoro_session(
    session_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete a pomodoro session"""
    service = PomodoroSessionService(db)
    success = service.delete_session(current_user.id, session_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Pomodoro session not found"
        )
    return {"message": "Pomodoro session deleted successfully"}
